engine/src/test/java/org/hibernate/validator/test/internal/constraintvalidators/hv/URLValidatorTest.java

```diff
@@ -290,10 +290,91 @@ private void runUrlContainerValidation(Validator validator, URLContainer contain
+		assertTrue( validator.isValid( "http://j.mp", null ) );
+		assertTrue( validator.isValid( "ftp://foo.bar/baz", null ) );
+		assertTrue( validator.isValid( "http://foo.bar/?q=Test%20URL-encoded%20stuff", null ) );
+		assertTrue( validator.isValid( "http://مثال.إختبار", null ) );
+		assertTrue( validator.isValid( "http://例子.测试", null ) );
+		assertTrue( validator.isValid( "http://उदाहरण.परीक्षा", null ) );
+		assertTrue( validator.isValid( "http://-.~_!$&'()*+,;=:%40:80%2f::::::@example.com", null ) );
+		assertTrue( validator.isValid( "http://1337.net", null ) );
+		assertTrue( validator.isValid( "http://a.b-c.de", null ) );
+		assertTrue( validator.isValid( "http://223.255.255.254", null ) );
+
+		// invalid urls:
 		assertFalse( validator.isValid( "http", null ) );
 		assertFalse( validator.isValid( "ftp//abc.de", null ) );
-		assertTrue( validator.isValid( "ftp://abc.de", null ) );
+		assertFalse( validator.isValid( "//", null ) );
+		assertFalse( validator.isValid( "//a", null ) );
+		assertFalse( validator.isValid( "///", null ) );
+		assertFalse( validator.isValid( "///a", null ) );
+		assertFalse( validator.isValid( "foo.com", null ) );
+		assertFalse( validator.isValid( ":// should fail", null ) );
+//		assertFalse( validator.isValid( "rdar://1234", null ) );
+//		assertFalse( validator.isValid( "ftps://foo.bar/", null ) );
+//		assertFalse( validator.isValid( "h://test", null ) );
+
+		if ( validator instanceof RegexpURLValidator ) {
+//			assertFalse( validator.isValid( "http://", null ) );
+//			assertFalse( validator.isValid( "http://.", null ) );
+//			assertFalse( validator.isValid( "http://..", null ) );
+//			assertFalse( validator.isValid( "http://../", null ) );
+//			assertFalse( validator.isValid( "http://?", null ) );
+//			assertFalse( validator.isValid( "http://??", null ) );
+//			assertFalse( validator.isValid( "http://??/", null ) );
+//			assertFalse( validator.isValid( "http://#", null ) );
+//			assertFalse( validator.isValid( "http://##", null ) );
+//			assertFalse( validator.isValid( "http://##/", null ) );
+			assertFalse( validator.isValid( "http://foo.bar?q=Spaces should be encoded", null ) );
+//			assertFalse( validator.isValid( "http:///a", null ) );
+			assertFalse( validator.isValid( "http:// shouldfail.com", null ) );
+			assertFalse( validator.isValid( "http://foo.bar/foo(bar)baz quux", null ) );
+//			assertFalse( validator.isValid( "http://-error-.invalid/", null ) );
+//			assertFalse( validator.isValid( "http://a.b--c.de/", null ) );
+//			assertFalse( validator.isValid( "http://-a.b.co", null ) );
+//			assertFalse( validator.isValid( "http://a.b-.co", null ) );
+//			assertFalse( validator.isValid( "http://0.0.0.0", null ) );
+//			assertFalse( validator.isValid( "http://10.1.1.0", null ) );
+//			assertFalse( validator.isValid( "http://10.1.1.255", null ) );
+//			assertFalse( validator.isValid( "http://224.1.1.1", null ) );
```

Comment:
here's some explanation from the author of that regexp:

> I have added simple network ranges validation, the rules I used are:
> - valid range 1.0.0.0 - 223.255.255.255, network adresses above and including 224.0.0.0 are reserved addresses
> - first and last IP address of each class is excluded since they are used as network broadcast addresses
> since I don't think this is worth implementing completely in a regular expression, a following pass should exclude the Intranet address space:
> 10.0.0.0 - 10.255.255.255
> 172.16.0.0 - 172.31.255.255
> 192.168.0.0 - 192.168.255.255
> the loopback and the automatic configuration address space:
> 127.0.0.0 - 127.255.255.255
> 169.254.0.0 - 169.254.255.255
> while the local, multicast and and the reserved address spaces:
> 0.0.0.0 - 0.255.255.255 (SPECIAL-IPV4-LOCAL-ID-IANA-RESERVED)
> 224.0.0.0 - 239.255.255 (MCAST-NET)
> 240.0.0.0 - 255.255.255.255 (SPECIAL-IPV4-FUTURE-USE-IANA-RESERVED)
> should already be excluded by the above regular expression.

I've extracted the validation used in EmailValidator as @gsmet suggested in a few comments above, and left commented out those cases that still do not pass.